test/upgrade_test.yaml

```diff
@@ -0,0 +1,134 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: upgrade-test-
+spec:
+  entrypoint: upgrade-test
+  volumes:
+  - name: gcp-credentials
+    secret:
+      secretName: user-gcp-sa
+  arguments:
+    parameters:
+    - name: image-build-context-gcs-uri
+    - name: target-image-prefix
+    - name: test-results-gcs-dir
+    - name: test-results-file-name
+    - name: cleanup
+    - name: upgrade-test-image-suffix
+      value: upgrade_test
+    - name: cluster-type
+      value: gke
+    - name: namespace
+      value: kubeflow
+  templates:
+  - name: upgrade-test
+    inputs:
+      parameters:
+      - name: target-image-prefix
+      - name: test-results-gcs-dir
+      - name: test-results-file-name
+      - name: cleanup
+      - name: upgrade-test-image-suffix
+      - name: namespace
+    steps:
+    - - name: build-upgrade-test-image
+        template: build-image
+        arguments:
+          parameters:
+          - name: docker-path
+            value: .
+          - name: docker-file
+            value: test/backend-test/Dockerfile
+          - name: image-name
+            value: "{{inputs.parameters.target-image-prefix}}{{inputs.parameters.upgrade-test-image-suffix}}"
+    - - name: run-upgrade-tests
+        template: run-upgrade-tests
+        arguments:
+          parameters:
+          - name: test-results-gcs-dir
+            value: "{{inputs.parameters.test-results-gcs-dir}}"
+          - name: test-results-file-name
+            value: "{{inputs.parameters.test-results-file-name}}"
+          - name: cleanup
+            value: "{{inputs.parameters.cleanup}}"
+          - name: upgrade-test-image
+            value: "{{inputs.parameters.target-image-prefix}}{{inputs.parameters.upgrade-test-image-suffix}}"
+
+  # Build and push image
+  - name: build-image
+    inputs:
+      parameters:
+      # GCS URI prefix pointing to a .tar.gz archive of Docker build context
+      - name: image-build-context-gcs-uri
+        value: "{{workflow.parameters.image-build-context-gcs-uri}}"
+      # The relative code path to the Dockerfile
+      - name: docker-path
+      # Name of the Docker file to use. "Dockerfile" by default
+      - name: docker-file
+        value: Dockerfile
+      - name: image-name
+    outputs:
+      parameters:
+        - name: strict-image-name
```

Comment:
why output strict-image-name? it does not seem to be used anywhere else.

Reply:
It's copy-pasted template from other test workflows.
I've added this output so that it's possible to reference the exact image that was built by the build-image task.
This way the caller does not need to care about the image name as the strict-image-mane is going to be unique.

Reply:
It would be better to pass this output to the downstream tasks.